runtime-sdk/src/crypto/random.rs

```diff
@@ -0,0 +1,94 @@
+use merlin::{Transcript, TranscriptRng};
+use rand_core::{CryptoRng, RngCore};
+use schnorrkel::keys::{ExpansionMode, MiniSecretKey};
+
+use crate::context::Context;
+
+pub struct Rng {
+    state: Option<RngState>, // `None` if the RNG is inoperable.
+}
+
+struct RngState {
+    transcript: Transcript,
+    rng: TranscriptRng,
+}
+
+impl Rng {
+    /// Creates a new RNG, potentially seeded using the provided `ctx`.
+    /// This should only be called once per top-level context.
+    pub fn new<C: Context + ?Sized>(ctx: &C) -> Self {
+        let km = match ctx.key_manager() {
+            Some(km) => km,
+            None => return Self { state: None },
+        };
+        let round_header_hash = ctx.runtime_header().encoded_hash();
+        let key_id = crate::keymanager::get_key_pair_id([
+            b"oasis-runtime-sdk/crypto: random_bytes".as_slice(),
+            &[ctx.mode() as u8],
+            round_header_hash.as_ref(),
+        ]);
+        let km_kp = match km
+            .get_or_create_keys(key_id)
+            .ok()
+            .map(|ks| ks.input_keypair)
+        {
+            Some(kp) => kp,
+            None => {
+                return Self { state: None };
```

Comment:
why wouldn't one entity aborting not also be a discrepancy, just of a different kind?

Reply:
Yeah this is exactly the thing.

If you just silently compute a different result such a discrepancy is subject to slashing for incorrect results.

But if the compute node knows that there was a problem while executing a batch it will instead indicate an explicit failure. Which may lead to availability slashing down the road (but this depends also whether other nodes were able to perform or not).